### PR TITLE
Voucher codes component

### PR DESCRIPTION
**Pull request by poulch**
Voucher codes component with generating and paginating voucher codes on the client side hidden behind a feature flag.
Run this code in browser console to set the enable feature flag

```
localStorage.setItem('FF_voucher_codes', '{"enabled":true,"payload":"default"}')
```

AC:
- As user I should see voucher codes only when feature flag is enable
- As user I should be able to generate voucher codes manually by providing my own voucher code
- As user I should be able to generate voucher codes automatically by providing quantity and prefix
- As use I should be able to paginate voucher codes
- Voucher codes are client only and will not be saved in API

### Pull Request Checklist

1. [ ] This code contains UI changes
2. [ ] All visible strings are translated with proper context including data-formatting
3. [ ] Attributes `data-test-id` are added for new elements
4. [ ] The changes are tested in Chrome/Firefox/Safari browsers and in light/dark mode
5. [ ] Your code works with the latest stable version of the core
6. [ ] I added changesets and [read good practices](../.changeset/README.md)

### Test environment config

API_URI=https://master.staging.saleor.cloud/graphql/
APPS_MARKETPLACE_API_URI=https://apps.staging.saleor.io/api/v2/saleor-apps

### Do you want to run more stable tests?

To run all tests, just select the stable checkbox. To speed up tests, increase the number of containers. Tests will be re-run only when the "run e2e" label is added.

1. [x] stable
2. [ ] app
3. [ ] attribute
4. [ ] category
5. [ ] collection
6. [ ] customer
7. [ ] giftCard
8. [ ] homePage
9. [ ] login
10. [ ] menuNavigation
11. [ ] navigation
12. [ ] orders
13. [ ] pages
14. [ ] payments
15. [ ] permissions
16. [ ] plugins
17. [ ] productType
18. [ ] products
19. [ ] sales
20. [ ] shipping
21. [ ] translations
22. [ ] variants
23. [x] vouchers

CONTAINERS=6

### CODE REVIEW

File: src/discounts/components/VoucherCodes/VoucherCodes.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  selectedCodesIds,
+  onMultiCodesGenerate,
+  onCustomCodeGenerate,
+  onDeleteCodes,
+  voucherCodesPagination,
+  ...datagridProps
+}: VoucherCodesProps) => {
+  const { pageInfo, ...paginationValues } = voucherCodesPagination;
+  const [openModal, setOpenModal] = useState<
+    "single-codes" | "multiple-codes" | "delete-codes" | null
```
Comment (Droniu) [inline src/discounts/components/VoucherCodes/VoucherCodes.tsx:42]: **suggestion**
```suggestion
  VoucherCodesUrlDialog | null
```

File: src/discounts/components/VoucherCodesAddButton/VoucherCodesAddButton.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    onMultiCodesGenerate();
+    setSubMenuOpen(false);
+  }, []);
+
+  const handleManualCodeGenerate = useCallback(() => {
+    onSingleCodesGenerate();
+    setSubMenuOpen(false);
+  }, []);
+
+  const subMenuItems = useMemo(
```
Comment (Droniu) [inline src/discounts/components/VoucherCodesAddButton/VoucherCodesAddButton.tsx:36]: **nitpick** I'd add explicit type here

File: src/discounts/components/VoucherCodesDatagrid/VoucherCodesDatagrid.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import { messages } from "./messages";
+import { VoucherCode } from "./types";
+
+export interface VoucherCodesDatagridProps {
+  codes: VoucherCode[];
+  loading: boolean;
+  disabled?: boolean;
+  onSelectVoucherCodesIds: (rows: number[], clearSelection: () => void) => void;
+  settings: any;
+  onSettingsChange: any;
```
Comment (Droniu) [inline src/discounts/components/VoucherCodesDatagrid/VoucherCodesDatagrid.tsx:25]: 👀

File: src/discounts/components/VoucherCodesDatagrid/datagrid.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    switch (columnId) {
+      case "code":
+        return readonlyTextCell(rowData?.code ?? "", false);
+      case "usage":
+        return readonlyTextCell(
+          rowData?.used?.toString() ?? PLACEHOLDER,
+          false,
+        );
+      case "status":
+        return readonlyTextCell(getStatus(rowData?.isActive, intl), false);
```
Comment (Droniu) [inline src/discounts/components/VoucherCodesDatagrid/datagrid.ts:47]: **question:** we have a status cell, perhaps we could use it here?

File: src/discounts/components/VoucherCodesDeleteDialog/VoucherCodesDeleteDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+export const VoucherCodesDeleteDialog = ({
+  open,
+  onClose,
+  onDelete,
+}: VoucherCodesDeleteDialogProps) => {
+  const intl = useIntl();
+
+  return (
+    <Modal open={open} onChange={onClose}>
```
Comment (Droniu) [inline src/discounts/components/VoucherCodesDeleteDialog/VoucherCodesDeleteDialog.tsx:21]: **suggestion**: can we use`@dashboard/components/Modal` here?

File: src/discounts/components/VoucherCodesDeleteDialog/VoucherCodesDeleteDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+          backgroundColor="surfaceNeutralPlain"
+          boxShadow="modal"
+          __left="50%"
+          __top="50%"
+          position="fixed"
+          __maxWidth="400px"
+          __transform="translate(-50%, -50%)"
+        >
+          <DashboardCard>
+            <DashboardCard.Title>Delete voucher code</DashboardCard.Title>
```
Comment (Droniu) [inline src/discounts/components/VoucherCodesDeleteDialog/VoucherCodesDeleteDialog.tsx:33]: **issue (blocking)** untranslated string

File: src/discounts/components/VoucherLimits/VoucherLimits.tsx [hunk 7]
```diff
@@ -110,6 +113,15 @@ const VoucherLimits = ({
           name={"onlyForStaff" as keyof VoucherDetailsPageFormData}
           onChange={onChange}
         />
+        {voucherCodesFlag.enabled && (
+          <ControlledCheckbox
+            testId="single-use"
+            checked={data.singleUse}
+            label={intl.formatMessage(messages.singleUse)}
+            name={"singleUse" as keyof VoucherDetailsPageFormData}
```
Comment (Droniu) [inline src/discounts/components/VoucherLimits/VoucherLimits.tsx:121]: **nitpick**
```suggestion
            name={"singleUse" satisfies keyof VoucherDetailsPageFormData}
```

File: src/discounts/views/VoucherDetails/hooks/useVoucherCodes.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    handleDeleteAddedVoucherCodes(selectedVoucherCodesIds);
+  };
+
+  return {
+    voucherCodes: clientVoucherCodes,
+    addedVoucherCodes,
+    voucherCodesLoading: false,
+    voucherCodesPagination: clientVoucherCodesPagination,
+    voucherCodesSettings,
+    updateVoucherCodesListSettings: (key: any, value: any) => {
```
Comment (Droniu) [inline src/discounts/views/VoucherDetails/hooks/useVoucherCodes.ts:42]: 👀

File: src/discounts/views/VoucherDetails/hooks/useVoucherCodesClient.test.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
+import { act } from "@testing-library/react";
+import { renderHook } from "@testing-library/react-hooks";
+
+import { useVoucherCodesClient } from "./useVoucherCodesClient";
+
+jest.mock("uuid", () => ({ v4: () => "123456789" }));
```
Comment (Droniu) [inline src/discounts/views/VoucherDetails/hooks/useVoucherCodesClient.test.ts:6]: **nitpick** - I think it's better to use actual uuid v4 here
```suggestion
jest.mock("uuid", () => ({ v4: () => "68276b31-3b41-4004-acd6-bad8c36d524f" }));
```

**Review by Droniu — CHANGES_REQUESTED** [review]

File: src/hooks/useLocalPaginator.ts [hunk 10]
```diff
@@ -88,13 +88,20 @@ export function useSectionLocalPaginationState(
   ];
 }
 
+export interface LocalPagination {
+  loadNextPage: () => void;
+  loadPreviousPage: () => void;
+  paginatorType: "click";
```
[the remark this answers is not in the document]
Comment (poulch) [reply src/hooks/useLocalPaginator.ts:94]: Not in this case IMO, this type only reflect what hook return so I can use this type.

File: src/discounts/components/VoucherCodesGenerateDialog/VoucherCodesGenerateDialog.tsx [hunk 11]
```diff
@@ -86,7 +76,7 @@ export const VoucherCodesGenerateDialog = ({
             label={intl.formatMessage(messages.codeQuantity, {
               maxCodes: MAX_VOUCHER_CODES,
             })}
-            onKeyDown={handleKeyDown}
+            // onKeyDown={handleKeyDown}
```
Comment (Droniu) [inline src/discounts/components/VoucherCodesGenerateDialog/VoucherCodesGenerateDialog.tsx:79]: **nitpick**: can be deleted

**Review by Droniu — APPROVED** [review]

**Pull request merged**